File: image_crop_compress/image_crop_compress_android/android/src/main/kotlin/com/imageprocessor/image_crop_compress_android/ConvertEngine.kt
package com.imageprocessor.image_crop_compress_android

import android.content.Context
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import java.io.File
import java.io.FileOutputStream
import java.util.UUID

object ConvertEngine {
    fun handle(call: MethodCall, result: MethodChannel.Result, context: Context) {
        val sourcePath = call.argument<String>("sourcePath")
        val targetFormatString = call.argument<String>("targetFormat")
        val quality = call.argument<Int>("quality") ?: 100

        if (sourcePath == null || targetFormatString == null) {
            result.error("INVALID_ARGUMENT", "sourcePath and targetFormat are required", null)
            return
        }

        val compressFormat = when (targetFormatString.lowercase()) {
            "png" -> Bitmap.CompressFormat.PNG
            "webp" -> if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.R)
                Bitmap.CompressFormat.WEBP_LOSSY
            else
                @Suppress("DEPRECATION") Bitmap.CompressFormat.WEBP
            "heif" -> if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.R)
                Bitmap.CompressFormat.WEBP_LOSSY // Fallback since native HEIF encoding is complex
            else Bitmap.CompressFormat.JPEG // Fallback
            else -> Bitmap.CompressFormat.JPEG
        }

        CoroutineScope(Dispatchers.IO).launch {
            try {
                val bitmap = BitmapFactory.decodeFile(sourcePath)
                    ?: throw Exception("Failed to decode image at $sourcePath")

                val tempDir = File(context.cacheDir, "image_crop_compress")
                if (!tempDir.exists()) tempDir.mkdirs()
                val outputFile = File(tempDir, "convert_${UUID.randomUUID()}.$targetFormatString")

                FileOutputStream(outputFile).use { out ->
                    bitmap.compress(compressFormat, quality, out)
                }

                val sizeInBytes = outputFile.length()
                val width = bitmap.width
                val height = bitmap.height
                bitmap.recycle()

                val mimeType = when (compressFormat) {
                    Bitmap.CompressFormat.PNG -> "image/png"
                    Bitmap.CompressFormat.JPEG -> "image/jpeg"
                    else -> "image/webp"
                }

                withContext(Dispatchers.Main) {
                    result.success(mapOf(
                        "path" to outputFile.absolutePath,
                        "width" to width,
                        "height" to height,
                        "sizeInBytes" to sizeInBytes,
                        "extension" to targetFormatString,
                        "mimeType" to mimeType
                    ))
                }
            } catch (e: Exception) {
                withContext(Dispatchers.Main) {
                    result.error("CONVERT_ERROR", e.message, null)
                }
            }
        }
    }
}

File: packages/image_crop_compress/image_crop_compress_ios/ios/image_crop_compress_ios/Sources/image_crop_compress_ios/ConvertEngine.swift
import Flutter
import UIKit
import CoreGraphics

public class ConvertEngine {
    public static func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let sourcePath = args["sourcePath"] as? String,
              let targetFormat = args["targetFormat"] as? String else {
            result(FlutterError(code: "INVALID_ARGUMENT", message: "sourcePath and targetFormat are required", details: nil))
            return
        }
        
        let quality = args["quality"] as? Int ?? 100
        
        DispatchQueue.global(qos: .userInitiated).async {
            guard let image = UIImage(contentsOfFile: sourcePath) else {
                DispatchQueue.main.async {
                    result(FlutterError(code: "DECODE_ERROR", message: "Failed to decode image at \(sourcePath)", details: nil))
                }
                return
            }
            
            let formatStr = targetFormat.lowercased()
            
            let tempDir = FileManager.default.temporaryDirectory.appendingPathComponent("image_crop_compress")
            do {
                try FileManager.default.createDirectory(at: tempDir, withIntermediateDirectories: true, attributes: nil)
            } catch {
                DispatchQueue.main.async {
                    result(FlutterError(code: "IO_ERROR", message: "Failed to create temp directory", details: nil))
                }
                return
            }
            
            let fileName = "convert_\(UUID().uuidString).\(formatStr)"
            let fileURL = tempDir.appendingPathComponent(fileName)
            
            var data: Data?
            var mimeType = "image/jpeg"
            
            if formatStr == "png" {
                data = image.pngData()
                mimeType = "image/png"
            } else if formatStr == "webp" {
                // Future optimization: implement native webp encoding or rely on flutter plugins
                // For now, fallback to JPEG on iOS unless iOS 14+ native WebP encoding is hooked up
                data = image.jpegData(compressionQuality: CGFloat(quality) / 100.0)
                mimeType = "image/webp" 
            } else if formatStr == "heif" {
                // CoreImage HEIF support can be added. 
                // For now, fallback to JPEG.
                data = image.jpegData(compressionQuality: CGFloat(quality) / 100.0)
                mimeType = "image/heif"
            } else {
                data = image.jpegData(compressionQuality: CGFloat(quality) / 100.0)
                mimeType = "image/jpeg"
            }
            
            guard let finalData = data else {
                DispatchQueue.main.async {
                    result(FlutterError(code: "ENCODE_ERROR", message: "Failed to encode image", details: nil))
                }
                return
            }
            
            do {
                try finalData.write(to: fileURL)
                let sizeInBytes = try fileURL.resourceValues(forKeys: [.fileSizeKey]).fileSize ?? finalData.count
                
                DispatchQueue.main.async {
                    result([
                        "path": fileURL.path,
                        "width": Int(image.size.width),
                        "height": Int(image.size.height),
                        "sizeInBytes": sizeInBytes,
                        "extension": formatStr,
                        "mimeType": mimeType
                    ])
                }
            } catch {
                DispatchQueue.main.async {
                    result(FlutterError(code: "SAVE_ERROR", message: error.localizedDescription, details: nil))
                }
            }
        }
    }
}
